build/configure.js

```diff
@@ -0,0 +1,156 @@
+'use strict';
+
+var utils = require('utils');
+// DependencyGraph is to manage build backend file, including generate and
+// execute.
+var DependencyGraph = require('./configure/dependencyGraph');
+// BuildConfig is to manage generate all necessary ENV for each build backend.
+var BuildConfig = require('./configure/build_config');
+// AppConfigureStep is per app configure step generator.
+var AppConfigureStep = require('./configure/app_configure_step');
+
+// Default building steps before app's own building steps.
+const PRE_APP_BUILD_STEPS = [
+  'svoperapps',
+  'webapp-manifests',
+  'contacts-import-services',
+  'search-provider',
+  'keyboard-layouts',
+  'preferences',
+  'settings',
+  'webapp-shared',
+  'copy-common-files'
+];
+
+/**
+ * ConfigureStep would generate all the data for Buildbackend. Currently would
+ * contain PRE_APP_BUILD_STEPS and each app's steps.
+ * @param {object} options - build_config from main Makefile.
+ * @constructor
+ */
+var ConfigureStep = function(options) {
+  this.options = options;
+};
+
+ConfigureStep.prototype = {
+  /**
+   * Configure start from here.
+   */
+  start: function() {
+    var commonMkPath = utils.joinPath(this.options.GAIA_DIR, 'all.mk');
+    this.mainMake = new DependencyGraph(commonMkPath);
+
+    // TODO: we should be able to detect whether to regenerate makefile.
+    this.preAppConfig();
+    this.postAppConfig();
+    // build_config.in has contained all necessary ENV data for the generated
+    // makefile.
+    this.mainMake.genBackend('-include build_config.in\n');
+
+    // Execute the generated makefile directly.
+    this.mainMake.executeBackend();
+  },
+
+  /**
+   * Generate config data from PRE_APP_BUILD_STEPS.
+   */
+  preAppConfig: function() {
+    this._stepIndex = this.options.STAGE_DIR;
+    PRE_APP_BUILD_STEPS.forEach(function(step) {
+      // We don't generate setting.js if only build one app.
+      if (this.options.BUILD_APP_NAME !== '*' && step === 'settings') {
```

Comment:
In my concern, putting if-else logic directly will make it difficult for maintenance. If there are other modules that need to do same thing here, it could be more complicated.